src/core/control/xojfile/LoadHandler.cpp

```diff
@@ -1,389 +1,243 @@
+    this->fileVersion = 0;
+    this->minimalFileVersion = 0;
+
     this->zipFp = nullptr;
-    this->zipContentFile = nullptr;
-    this->gzFp = nullptr;
     this->isGzFile = false;
-    this->error = nullptr;
-    this->attributeNames = nullptr;
-    this->attributeValues = nullptr;
-    this->elementName = nullptr;
-    this->pdfFilenameParsed = false;
-    this->attachedPdfMissing = false;
+    this->xmlContentStream.reset();
 
-    this->page = nullptr;
-    this->layer = nullptr;
-    this->stroke = nullptr;
-    this->image = nullptr;
-    this->teximage = nullptr;
-    this->text = nullptr;
     this->pages.clear();
+    this->audioFiles.clear();
+
+    this->page.reset();
+    this->layer.reset();
+    this->stroke.reset();
+    this->image.reset();
+    this->teximage.reset();
+    this->text.reset();
+}
+
+auto LoadHandler::hasErrorMessages() const -> bool { return !this->errorMessages.empty(); }
 
-    if (this->audioFiles) {
-        g_hash_table_unref(this->audioFiles);
+auto LoadHandler::getErrorMessages() const -> std::string {
+    std::string messages;
+    for (const auto& error: this->errorMessages) {
+        messages += '\n' + error;
     }
-    this->audioFiles = g_hash_table_new(g_str_hash, g_str_equal);
+    return messages;
 }
 
-auto LoadHandler::getLastError() -> string { return this->lastError; }
-
 auto LoadHandler::isAttachedPdfMissing() const -> bool { return this->attachedPdfMissing; }
 
-auto LoadHandler::getMissingPdfFilename() const -> string { return this->pdfMissing; }
-
-auto LoadHandler::openFile(fs::path const& filepath) -> bool {
-    this->filepath = filepath;
-    int zipError = 0;
-    this->zipFp = zip_open(filepath.u8string().c_str(), ZIP_RDONLY, &zipError);
-
-    // Check if the file is actually an old XOPP-File and open it
-    if (!this->zipFp && zipError == ZIP_ER_NOZIP) {
-        this->gzFp = GzUtil::openPath(filepath, "r");
-        this->isGzFile = true;
-    }
-
-    if (this->zipFp && !this->isGzFile) {
-        // Check the mimetype
-        zip_file_t* mimetypeFp = zip_fopen(this->zipFp, "mimetype", 0);
-        if (!mimetypeFp) {
-            this->lastError = zip_error_strerror(zip_get_error(zipFp));
-            this->lastError =
-                    FS(_F("The file is no valid .xopp file (Mimetype missing): \"{1}\"") % filepath.u8string());
-            return false;
-        }
-        char mimetype[MAX_MIMETYPE_LENGTH + 1] = {};
-        // read the mimetype and a few more bytes to make sure we do not only read a subset
-        zip_fread(mimetypeFp, mimetype, MAX_MIMETYPE_LENGTH);
-        if (!strcmp(mimetype, "application/xournal++")) {
-            zip_fclose(mimetypeFp);
-            this->lastError = FS(_F("The file is no valid .xopp file (Mimetype wrong): \"{1}\"") % filepath.u8string());
-            return false;
-        }
-        zip_fclose(mimetypeFp);
+auto LoadHandler::getMissingPdfFilename() const -> fs::path { return this->missingPdf; }
```

Comment:
I understand that nothing would break here, but isn't it generally considered bad practice to return references to class members outside the class? (because of the risk of dangling reference)

Reply:
It's the other way round. Most usages are read only, no reason to do a copy on each call. When you need a copy for later usage, you can still do that.